transit.md

```diff
@@ -217,3 +272,108 @@ both sides have their own nonce.
 For each received record, the nonce must be checked to be equal to the expected value,
 which has to be tracked. Thus, there are 4 nonces in total, two per direction which are
 kept in sync.
+
+## On finding a direct connection
+
+Because it is easy to get things wrong. This section only applies to the `direct-tcp-v1`
+hint. The following section is closely tied to the operating system's sockets API, and the
+the verbs *bind*, *listen*, *accept* and *connect* will be used accordingly. Note that this
+is more a guide than a specification, since it handles implementation details that should
+all mostly be compatible with each other (and when not, failure is not critical).
+
+### The easy way
+
+Each side binds a new TCP socket, enumerates all local non-loopback IP addresses and exchanges
+port+IPs with the peer. Then, both sides listen on their socket while simultaneously trying to
+connect to the peer. As described above, once a handshake is successful, all other attempts are
+aborted and the resources closed.
+
+### Punching holes through firewalls
+
+To go through firewalls, a different approach must be used instead: first, both sides again need
+to bind a socket and exchange hints. This time however, the `SO_REUSEADDR` option must be set.
+Then, both sides try to connect to each of the peer's IPs. For every connection attempt, they bind
+a new socket on the same port again. Also you read that right, there are no `listen` or `accept`
+calls involved.
+
+This works only with firewalls that silently filter packets and which don't send back any RST packets.
+Unfortunately, this is the default behavior of the kernel when no firewall is active (or the firewall allowed)
+the packets. This means that the firewall hole punching method does *not* work when *no* firewall is
+active (Fun fact: if one side has a firewall and the other side doesn't then connection attempts will
+work or fail depending on which side sent their packets first—I'm really glad for you that you didn't
+have to debug this …). Thus, **this method can only be an addition to the "easy" way** specified above.
+Sadly, this means that the number of hints – and thus, attempted connections – is doubled. If you manage
+to listen and accept on the same port so that only one set of hints needs to be sent out, please contact
+`piegames` with your findings. Also read [this paper](https://bford.info/pub/net/p2pnat/) (Python code
+[here](https://github.com/dwoz/python-nat-hole-punching/)) for more details on the subject.
+
+### Traversing NAT
+
+To additionally traverse through a potential NAT, clients need to query their external IP address
+from a STUN server\*. Because of how NATs work, the socket must again be bound with `SO_REUSEADDR`
```

Comment:
This might benefit from discussing symmetric vs cone etc etc NATs, or pointing at another resource for this .. there's of course a whole bunch of hole-punching research over the last two decades and we can't possibly be comprehensive here. (Specifically, I don't believe this technique will work for "restricted" NATs that require at least one outgoing packet to the destination .. and some require the right port too). Of course, we don't "need" any particular technique to work since they're "just" hints, but ...

Reply:
I try to keep this paragraph as "just do this and it'll work most of the time", since this topic is already complex enough. But I'll add a link to the paper that helped me implementing – although I didn't use their algorithm (couldn't get it to work), they go really in depth on all the NAT types and different scenarios and edge cases.